### DISCUSSION TITLE
Warp

### DISCUSSION THREAD

**Discussion by ronanyeah** [rwf2/multer#39] [Q&A] [answered]
Is it possible to pass the output of this warp filter to multer?
https://docs.rs/warp/0.3.1/warp/filters/body/fn.stream.html

I can't get it to work.
```
async fn handler<S, B>(form: S) -> Result<impl warp::Reply, warp::Rejection>
where
    S: Stream<Item = Result<B, warp::Error>>,
    S: StreamExt,
    B: warp::Buf,
{
```
From what I can gather, `warp::Buf` needs to satisfy `Into<Bytes>`?

**Comment by rousan** [comment]
@SergioBenitez

**Comment by SergioBenitez** [comment, marked as answer]
Yes, but it's not clear to me from warp's sparse documentation how to do so efficiently and correctly. The gist is that you need to convert the `impl Buf` (note that `Buf` is a trait, not a concrete type) into something that implements `Into<Bytes>`. Here are some ideas:

  * Map the stream into a stream of `Bytes` using [copy_to_bytes](https://docs.rs/bytes/1.0.1/bytes/buf/trait.Buf.html#method.copy_to_bytes) on the `impl Buf`.
  * Use the [bytes](https://docs.rs/warp/0.3.1/warp/filters/body/fn.bytes.html) filter instead of `stream`.
  * The use [form](https://docs.rs/warp/0.3.1/warp/filters/multipart/fn.form.html) filter instead of `multer`.

All of these do much more copying and allocating than one would hope, but it's not clear to be that warp allows anything better.

**Comment by ronanyeah** [comment]
I got #2 working, thanks a lot to both of you. Here's the code:

```
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let env = envy::from_env::<Env>().expect("Missing environment variables!");

    let routes = warp::post()
        .and(warp::path("new-listing"))
        .and(warp::filters::header::value("Content-Type"))
        .and(warp::filters::body::bytes())
        .and_then(handle_form);

    warp::serve(routes).run(([0, 0, 0, 0], env.port)).await;

    Ok(())
}

async fn handle_form(
    content_type: warp::http::header::HeaderValue,
    form: bytes::Bytes,
) -> Result<impl warp::Reply, warp::Rejection> {
    let stream = futures::stream::once(async move { Result::<bytes::Bytes, Infallible>::Ok(form) });

    let boundary = multer::parse_boundary(content_type.to_str().unwrap()).unwrap();

    let mut multipart = multer::Multipart::new(stream, boundary);

    while let Some(field) = multipart.next_field().await.unwrap() {
        println!("Field: {:?}", field.text().await)
    }

    let response = serde_json::json!({ "message": "OK" });
    Ok(warp::reply::json(&response))
}
```
